File: dynamic_zero_export/runtime/topology.ts
import { type DynamicZeroExportSiteConfig, type TopologyType } from '../schema/site-config.types';

export type DerivedTopology = {
  type: TopologyType;
  mode: 'single' | 'dual-separate' | 'dual-combined' | 'ambiguous';
  zones: Array<{ id: string; label: string; busSide: 'A' | 'B' | 'both' }>;
};

export function deriveTopology(config: DynamicZeroExportSiteConfig): DerivedTopology {
  if (config.topology.type === 'DUAL_BUS_COMBINED') {
    return {
      type: config.topology.type,
      mode: 'dual-combined',
      zones: [{ id: 'combined', label: 'Combined Zone', busSide: 'both' }],
    };
  }
  if (config.topology.type === 'DUAL_BUS_SEPARATE') {
    return {
      type: config.topology.type,
      mode: 'dual-separate',
      zones: [
        { id: 'bus_a', label: 'Bus A', busSide: 'A' },
        { id: 'bus_b', label: 'Bus B', busSide: 'B' },
      ],
    };
  }
  if (config.topology.type.startsWith('DUAL_BUS')) {
    return {
      type: config.topology.type,
      mode: config.topology.tieSignalPresent ? 'dual-separate' : 'ambiguous',
      zones: [
        { id: 'bus_a', label: 'Bus A', busSide: 'A' },
        { id: 'bus_b', label: 'Bus B', busSide: 'B' },
      ],
    };
  }
  return {
    type: config.topology.type,
    mode: 'single',
    zones: [{ id: 'site', label: 'Site', busSide: 'both' }],
  };
}

File: dynamic_zero_export/firmware_core/src/topology.cpp
#include "dzx/topology.hpp"

namespace dzx {

TopologyState deriveTopology(const DynamicZeroExportSiteConfig& cfg) {
  TopologyState state;
  state.type = cfg.topology.type;
  state.dualBus = cfg.topology.busCount == 2;
  state.ambiguous = false;

  if (cfg.topology.type == TopologyType::DualBusCombined) {
    state.mode = "dual-combined";
    state.controlZones = 1;
  } else if (cfg.topology.type == TopologyType::DualBusSeparate) {
    state.mode = "dual-separate";
    state.controlZones = 2;
  } else if (cfg.topology.type == TopologyType::SingleBusMultiGen) {
    state.mode = "single-multi-gen";
    state.controlZones = 1;
  } else if (cfg.topology.type == TopologyType::DualBus && cfg.topology.tieSignalPresent) {
    state.mode = "ambiguous";
    state.controlZones = 2;
    state.ambiguous = true;
  } else if (cfg.topology.type == TopologyType::DualBus) {
    state.mode = "dual";
    state.controlZones = 2;
  } else {
    state.mode = "single";
    state.controlZones = 1;
  }
  return state;
}

}  // namespace dzx

